lib/improver/nowcasting/accumulation.py

```diff
@@ -0,0 +1,270 @@
+        """
+        self.accumulation_units = accumulation_units
+
+    def __repr__(self):
+        """Represent the plugin instance as a string."""
+        result = '<Accumulation: accumulation_units={}>'
+        return result.format(self.accumulation_units)
+
+    @staticmethod
+    def sort_cubes_by_time(cubes):
+        """
+        Sort cubes in time ascending order (from earliest time to latest time).
+
+        Args:
+            cubes (iris.cube.CubeList):
+                A cubelist containing input precipitation rate cubes.
+        Returns:
+            cubes (iris.cube.CubeList):
+                The cubelist in ascending time order.
+            times (list):
+                A list of the validity times of the precipitation rate cubes in
+                integer seconds since 1970-01-01 00:00:00.
+        """
+        times = np.array([cube.coord('time').points[0] for cube in cubes])
+        time_sorted = np.argsort(times)
+        times = times[time_sorted]
+        cubes = iris.cube.CubeList(np.array(cubes)[time_sorted])
+        return cubes, times
+
+    def process(self, cubes):
+        """
+        Calculate period precipitation accumulations based upon precipitation
+        rate fields.
+
+        Args:
+            cubes (iris.cube.CubeList):
+                A cubelist containing input precipitation rate cubes.
+        Returns:
+            accumulation_cubes (iris.cube.CubeList):
+                A cubelist containing precipitation accumulation cubes where
+                the accumulation periods are determined by the intervals
+                between the precipitation rate cubes.
+        """
+        # Standardise inputs to expected units.
+        cubes = enforce_coordinate_units_and_dtypes(
+            cubes, ['time', 'forecast_reference_time', 'forecast_period'],
+            inplace=False)
+        enforce_diagnostic_units_and_dtypes(cubes)
+
+        # Sort cubes into time order and calculate intervals.
+        cubes, times = self.sort_cubes_by_time(cubes)
+        time_intervals = np.diff(times, axis=0)
+        accumulation_cubes = iris.cube.CubeList()
+
+        # Assumes rates cubes bookend the accumulations, e.g. there is one
+        # more rates cube than expected accumulation cubes.
+        iterator = zip(cubes[:-1], cubes[1:], time_intervals)
+
+        # Calculate accumulations and convert to desired units.
+        for start_cube, end_cube, time_interval in iterator:
+
+            # Use CubeCombiner to expand coordinate bounds.
+            cube_name = 'lwe_thickness_of_precipitation_amount'
+            accumulation_cube = CubeCombiner('add').process(
+                iris.cube.CubeList([start_cube, end_cube]), cube_name,
+                expanded_coord={'time': 'upper', 'forecast_period': 'upper'})
+            accumulation_cube.units = 'm'
+
+            # Calculate new data and insert into cube.
+            accumulation_data = start_cube.data * time_interval
```

Comment:
I thought the point of this plugin was to produce advected / interpolated accumulations?  Eg put in 2 rain rate fields 15 minutes apart, get advection components, interpolate to 1 minute frequency using those components and sum the result to get the 15 minute accumulation.  Am I missing something here?

Reply:
As discussed: we'll write down the designs we're envisaging and compare.